### ISSUE TITLE
Is there a good reason not to allow direct access to properties whose name starts with a digit?

### ISSUE THREAD

**Issue by mklement0** [PowerShell/PowerShell#6946]
Currently, you cannot access an object's property if that property's name happens to start with a (decimal) _digit_:

```powershell
$obj = [pscustomobject] @{ '1a' = '1a''s value' }
 
 # !! BREAKS, because the property name starts with a digit.
$obj.1a
```

The error message you get is:

```none
...
Missing property name after reference operator.
...
```

The current _workarounds_ are to either use _quoting_ or to _indirectly_ reference the property via a _variable_ containing the name:

```powershell
# OK: Quote the property name
$obj.'1a'
$obj."1a"

# OK: Access the property via  a *variable*
$propName = '1a'
$obj.$propName
```

<sup>Note: `$obj.{1a}`, as mentioned by @george-chakhidze below, happens to work too, but it is  an accidental by-product of script-block stringification.
</sup>

Is there a good reason _not_ to allow direct use of the literal property name?
In other words: why shouldn't `$obj.1a` work as-is?

(By contrast, there are property names that do invariably require quoting, such as
 `([pscustomobject] @{ 'a.b' = 'a.b''s value' }).'a.b'`)

Environment data
----------------

Written as of:

```powershell
PowerShell Core v6.1.0-preview.2
```

**Comment by BrucePay** [comment]
@george-chakhidze `$obj.{1a}` works because the string value for the scriptblock `[1a}` is `'1a'. So it's essentially equivalent to `$obj.'1a'`.

@mklement0 It's not allowed because it's sufficiently anomalous that it's probably a mistake. Cat's walking on keyboards should not produce valid programs. If you _really_ want to do it, then you have to make it explicit with quoting as in the case of this [StackOverflow question](https://stackoverflow.com/questions/50556895/powershell-properties-issues) (I assume this is what prompted the question?)

Now as it happens, `$foo.123` works fine. This is because we allow expressions of the LHS of `.` and, of course, `123` is a legal expression.

**Comment by mklement0** [comment]
@BrucePay:

Good to know that `$obj.{1a}` only works _accidentally_, as a by-product of script-block stringification (I've updated the original post).

> ... as in the case of this StackOverflow question. I assume this is what prompted the question?

Yes.
And it is tangible evidence that not only cats walking on keyboards produce such property names, especially in the world of JSON.

> It's not allowed because it's sufficiently anomalous that it's probably a mistake. 

If that is truly the intent - and I don't think it should be - please provide a more meaningful error message.

> Now as it happens, $foo.123 works fine. This is because we allow expressions of the LHS of . and, of course, 123 is a legal expression.

Yes, as it _happens_ - and I suggest not relying on things that _happen_.

Let me ask the opposite question: with respect to _property access_ on _objects_, what good reason is there to interpret what's to the right of `.` as anything other than a property _name_, i.e., a _string_?
(Hashtable keys are a different story).

**Comment by BrucePay** [comment]
@mklement0 The issue in the stackoverflow question was not difficulty in figuring out that quoting the anomalous property name was required. The problem was that there was a _typo_ in that name. Quoted or not made no difference since the property didn't exist.

> especially in the world of JSON.

Interestingly, JavaScript doesn't allow property names to start with a number either. You have you have to use indexing instead. That would seem to discourage the use of such names in JSON.

> If that is truly the intent - and I don't think it should be - please provide a more meaningful error message

As one of the people making that decision and the person who wrote the original code, I can say - yes, that is the intent. The error message is currently that the property is missing. I suppose we could change it to "missing or invalid".

> Yes, as it happens - and I suggest not relying on things that happen

Poor phrasing on my part. It's not an accident, it's the expected consequence of the explicit decision to allow expressions on the LHS of '.', which in turn was done, in part, to deal with scenarios where objects might have non-conforming property names. As is the case in the stackoverflow question.

**Comment by mklement0** [comment]
@BrucePay:

> The problem was that there was a _typo_ in that name.

Yes, the issue turned out to be a typo, but the fact remains that (a) digit-prefixed property names do occur, and (b) you do need explicitly quoting to access them, and that may be surprising.

> Interestingly, JavaScript doesn't allow property names

True, but JSON has long since outgrown use in the context of JavaScript _only_.

And, given that `var o = JSON.parse('{ "1a": "value" }')` works just fine even in JavaScript, you could similarly ask why JavaScript doesn't allow `o.1a` and instead requires `o['1a']` - but that's obviously a separate debate.

> The error message is currently that the property is missing. I suppose we could change it to "missing or invalid".

That would certainly help, though to be truly helpful it would have to say something along the lines of "when in doubt, _quote_ the name".

> which in turn was done, in part, to deal with scenarios where objects might have non-conforming property names. As is the case in the stackoverflow question.

Yet, typo aside, it is that very decision - treating what's to the right of `.` as an _expression_  ONLY - is the very thing _preventing_ the use of the non-conforming property name at hand.

Note that _falling back to `[string]` interpretation_ - if interpretation as an _expression_ doesn't work  - makes perfect sense, as _that is already happening for non-digit-prefixed barewords_ a.k.a regular property access:

```powershell
$co.foo # .foo is *implicitly* treated like .'foo', because foo by itself is not an expression
$co.1a # Why shouldn't this be treated like .'1a', if it can't be interpreted as an expression?
```

With object access, the advantage of being able to use an expression to the right of `.` is _solely_ to calculate a _property name_ - so that (ultimately) treating what's to the right as anything other than a string makes no sense.

**Comment by BrucePay** [comment]
> that (a) digit-prefixed property names do occur

Off the top of my head, I can't think of a single language that allows property names to start with a digit.

> Note that falling back to [string] interpretation - if interpretation as an expression doesn't work - makes perfect sense, as that is already happening for non-digit-prefixed barewords a.k.a regular property access

That's not what happens. The parser looks for either a propertyname token or a unary expression. If a propertyname token is found, it is used to look up the property. If an expression is found, that expression is evaluated, the result is converted to a string by calling `.ToString()` and that string is used to look up  the property.

> that (ultimately) treating what's to the right as anything other than a string makes no sense.

So you're saying that in `$x.abc+5` the property name should be `abc+5`?. That would make things rather awkward.

**Comment by vexx32** [comment]
I'm inclined to agree with that, really. It should default to string interpretation unless explicitly cast otherwise. However, it becomes tricky to distinguish there between what is intended to be a variable cast and what is simply part of a string. 

If you allow expressions to be recognised as strings in hashtable literals, then how is it supposed to treat type casts correctly?

```PowerShell
$TypeHash = @{
     [int]32 = "hello!"
}
# should this work?
$TypeHash.'[int]32'
# because it's not so simple to ensure that that's not the case if we just say 'parse everything as strings, unless... unless what, exactly?'
```

**Comment by vexx32** [comment]
I always wondered how `$Matches` managed to work almost like an array, but still function as a hashtable. Very interesting... I'll have to keep that in mind.

All said, I personally have no qualms with how it currently functions. It could be improved a little, perhaps, but it's pretty consistent as it is, and it does seem like changing the behaviour for, say, a specific use case, seems to be... Unwise, and liable to cause other problems. 

I don't really see an elegant solution here.

**Comment by mklement0** [comment]
@vexx32 

> I don't really see an elegant solution here.

Yes, unfortunately; arguably, the `.` syntactic sugar for hashtables keys should have applied _string-only_ interpretation from the start (and required differently typed keys to be accessed with `[...]` or even `.(...)`), but that ship has sailed.

>  say, a specific use case

It's more than a specific use case, but it is not common.

The best we can do under the circumstances is to provide a meaningful error message - see #6959
